### PR TITLE
Adding LM Head to Transfo-XL and first step to fixing problem with Adaptive Embeddings in TransfoXL

### LINKED ISSUE #3064: Add LM capabilities to TFTransfoXLLMHead
# 🚀 Feature request

It is currently not possible to generate language with the TFTransfoXLLMHeadModel because 
the `lm_head` is not implemented (see https://github.com/huggingface/transformers/blob/908fa43b543cf52a3238129624f502240725a6a6/src/transformers/modeling_tf_transfo_xl.py#L745)

Doing: 

```
from transformers import TFTransfoXLLMHeadModel
model = TFTransfoXLLMHeadModel.from_pretrained('transfo-xl-wt103') 
model(tf.convert_to_tensor([[1, 5, 8]])) 
```

currently leads to an error. 

## Motivation

Pytorch has TransfoXLLMHead model working - TF should has well.

## Your contribution

@LysandreJik , @thomwolf  I could implement it if needed.

### PR DESCRIPTION
**Pull request by patrickvonplaten**
This PR adds LM generation capabilitiies to the TF transfo-xl model. The integration tests for language generation pass, so generation from a pretrained model works now in TF as well. 

What does definitely not work yet is running the both PT and TF models with `self.sample_softmax > 0`: 
- Transfo-XL uses adaptive word embeddings -> the word embeddings ale broken down into 4 Embeddings of different shapes: `[20000, 1024], [20000, 1024], [160000, 64]` and `[67735, 16]` . When `self.sample_softmax > 0` though, it seems like the model expects the `normal` word embeddings with just a single weight matrix. When trying to tie the weights then as done in line 831 (see comment below), the logic breaks. 

This problem seems to be more complex though and I'd suggest to solve it in another PR (add possible have a call before to make things clear).

### CODE REVIEW

File: src/transformers/modeling_transfo_xl.py [hunk 1]
```diff
@@ -827,6 +827,7 @@ def tie_weights(self):
         # sampled softmax
         if self.sample_softmax > 0:
             if self.config.tie_weight:
+                # TODO: this does not make sense -> transformer.word_emb.weight does not exist
```
Comment (patrickvonplaten) [inline src/transformers/modeling_transfo_xl.py]: @LysandreJik @thomwolf . I didn't look deep enough into the paper and the code to understand everything fully, but even though all tests pass - there is a problem here. `self.out_layer.weight` is of dim `[267735, 1024]` where as `transformer.word_emd` is a list of differently shaped Embeddings -> not easy to copy weights here.

File: src/transformers/modeling_tf_transfo_xl.py [hunk 2]
```diff
@@ -745,13 +764,24 @@ def __init__(self, config):
         self.sample_softmax = config.sample_softmax
         # use sampled softmax
         if config.sample_softmax > 0:
-            raise NotImplementedError
+            self.out_layer = TFTransfoXLLMHead(config, self.transformer.word_emb.weight, name="out_layer")
```
Comment (patrickvonplaten) [inline src/transformers/modeling_tf_transfo_xl.py]: As is the case in PT: `self.transformer.word_emb.weight` does not work here, but added it for consistency.

File: src/transformers/modeling_tf_transfo_xl.py [hunk 3]
```diff
@@ -745,13 +764,24 @@ def __init__(self, config):
...
 
+    def get_output_embeddings(self):
+        """ Double-check if you are using adaptive softmax.
+        """
+        if self.sample_softmax > 0:
+            return self.out_layer
+        else:
+            if len(self.crit.out_layers) > 0:
+                return self.crit.out_layers[-1]
+            return None
```
Comment (patrickvonplaten) [inline src/transformers/modeling_tf_transfo_xl.py]: At initialization of `TFTransfoXLLMHead`, `self.crit.out_layers` is an empty list -> therefore only return last layer if it has one.

File: src/transformers/modeling_tf_transfo_xl_utilities.py [hunk 4]
```diff
@@ -176,3 +176,76 @@ def call(self, inputs, return_mean=True, training=False):
             self.add_metric(loss, name=self.name, aggregation="mean" if return_mean else "")
 
         return out
+
+
+# TODO: (PVP) add tests to compare to PT!
```
Comment (thomwolf) [inline src/transformers/modeling_tf_transfo_xl_utilities.py]: yes!

File: src/transformers/modeling_transfo_xl.py (hunk 1 above)
Comment (thomwolf) [reply src/transformers/modeling_transfo_xl.py]: I see.
We can remove the `sample_softmax` option if it makes the code cleaner.
The Transformer-XL codebase is definitely the most "researchy" one in the library and we've already had a few passes to remove dead code in it. Now could be a good time to remove the dead code related to Sampling Softmax (it's not used in pretrained models).

File: tests/test_modeling_tf_transfo_xl.py [hunk 5]
```diff
@@ -364,7 +364,7 @@ def test_lm_generate_transfo_xl_wt103(self):
                     0,
                 ]
             ],
-            dtype=tf.int31,
+            dtype=tf.int32,
```
Comment (thomwolf) [inline tests/test_modeling_tf_transfo_xl.py]: 😂

**Review by thomwolf — APPROVED** [review]
Ok good job.

I feel like we could remove all the dead code related to sampling softmax (see my comments)

File: tests/test_modeling_tf_transfo_xl.py (hunk 5 above)
Comment (patrickvonplaten) [reply tests/test_modeling_tf_transfo_xl.py]: Thought you can save memory like this :D

**Comment by patrickvonplaten** [comment]
> Ok good job.
> 
> I feel like we could remove all the dead code related to sampling softmax (see my comments)

Sound good, will do that then!

**Comment by patrickvonplaten** [comment]
Dead code is now removed. This removed a lot of code. To not re-invent the wheel the code is kept in the branch `add_sampling_and_training_to_transfo_xl_models` and documented by the feature request: #3310 , if someone wants to pick up implementing sample softmax again. 

This PR still adds language modeling capabilities to TF transfoXL.

**Review by thomwolf — APPROVED** [review]
Awesome!

**Review by LysandreJik — APPROVED** [review]
LGTM!
